basic_pacman/pacman_env.py

```diff
@@ -1,48 +1,165 @@
+        self._generate_pos('coin', NUM_COINS)
 
-        state = self.update_map()
+        state = self._update_map()
 
         return state
 
-    def __move(self):
+    def step(self, action: int) -> Tuple[np.ndarray, float, bool, str]:
+        """
+        This function formulates the (next_state, reward, done, info) values, thus moving the game one step forward.
+        :param action: chosen action 0/1/2/3 corresponding to the directions in order up/right/down/left)
+        :return:
+            next_state : ndarray
+                state of the environment after the action
+            reward : float
+                quantification of the state transition ###(To be implemented...)###
+            done : bool
+                True if the game is terminated, otherwise False
+            info : str
+                contains information about the game progress (scores acquired, steps)
+        """
+        next_state = self._get_next_state(action)
+        reward = self._get_reward(next_state)
+        done = self._check_done()
+        info = self._get_info()
+
+        return next_state, reward, done, info
+
+    def render(self) -> None:
+        """
+        This function creates a visualization of the game state.
+        :return: None
+        """
+        cv2.imshow("Pacman Environment", self.state)
+        cv2.waitKey(50)
+
+    def _get_info(self) -> str:
+        """
+        This function creates an 'info' string which can be used as a feedback from the environment itself.
+        :return: information about the game's current state
+        """
+        self._calculate_score()
+        info = "Points acquired: " + str(self.score) + "\nSteps: " + str(self.step_counter) + "/" + str(MAX_STEP)
+
+        return info
+
+    def _get_reward(self, state: np.ndarray) -> float:
+        """
+        This function realizes the reward function and formulates the value based on the state passed as an argument.
+        :param state: ndarray of the game state based on which the reward value is calculated
+        :return: reward value from which the agent is able to learn
+        """
+        # Only for RL purposes
+        # To be implemented later...
+        reward = 0.0
+
+        return reward
+
+    def _get_next_state(self, action: int) -> np.ndarray:
+        """
+        This function realizes the change in the game state according to the chosen action.
+        :param action: chosen action 0/1/2/3 corresponding to the directions in order up/right/down/left)
+        :return: state of the environment after the action
+        """
+        self._check_action_validity(action)
+        self._set_action(action)
+        self._move()
+        next_state = self._update_map()
+
+        return next_state
+
+    def _move(self) -> None:
+        """
+        This function modifies PacMan's position according to its facing direction.
+        :return: None
+        """
         if self.orientation == 0:
```

Comment:
Perhaps You can add STEP_SIZE, but the speed can also be changed with the step time (timer impl). It is up to you.

Reply:
Thanks for the feedback @NyesteAmbrusPal !
However, as the implementation of the Pacman environment is just like a grid world, it turned out for me, that using a step size not equal to 1 is not really effective (or actually not even optional) from the player's point of view.
Thus, as you have mentioned, for the sake of modifiable difficulty, the other way around is to accelerate the action selection process through timing.
Hope I have addressed your concern... :)